### PR TITLE
Sidgen improvements

### PR DESCRIPTION
**Pull request by abbra**

### CODE REVIEW

**Comment by flo-renaud** [comment]
### sidgen fix
Tested with a staged user, calling `/usr/libexec/ipa/oddjob/org.freeipa.server.config-enable-sid --add-sids`
Before the fix: 389ds error log contains
```
DEBUG - ipa-sidgen-postop - Trying to add SID for [uid=stageuser,cn=staged users,cn=accounts,cn=provisioning,dc=ipa,dc=test].
ERR - find_sid_for_ldap_entry - [file ipa_sidgen_common.c, line 483]: ID value too large.
ERR - do_work - [file ipa_sidgen_task.c, line 154]: Cannot add SID to existing entry.
DEBUG - ipa-sidgen-postop - do_work finished with [19].
ERR - sidgen_task_thread - [file ipa_sidgen_task.c, line 199]: Sidgen task finished [19].
```
After the fix:
```
DEBUG - ipa-sidgen-postop - Trying to add SID for [uid=stageuser,cn=staged users,cn=accounts,cn=provisioning,dc=ipa,dc=test].
DEBUG - ipa-sidgen-postop - Staged entry [uid=stageuser,cn=staged users,cn=accounts,cn=provisioning,dc=ipa,dc=test] does not have Posix IDs, nothing to do.
DEBUG - ipa-sidgen-postop - do_work finished with [0].
ERR - sidgen_task_thread - [file ipa_sidgen_task.c, line 199]: Sidgen task finished [0].
```

### canonical name
 I tested 2 scenarios:
- [x] Create a service, get a keytab for the service, use ldapmodify to remove the krbcanonicalname, try to kinit.
Before the fix:
```
# kinit -kt /tmp/svc.keytab MYSVC/server.ipa.test@IPA.TEST
kinit: Generic error (see e-text) while getting initial credentials
```
krb5kdc.log:
```
Jan 30 04:08:33 server.ipa.test krb5kdc[7795](info): AS_REQ (6 etypes {aes256-cts-hmac-sha384-192(20), aes128-cts-hmac-sha256-128(19), aes256-cts-hmac-sha1-96(18), aes128-cts-hmac-sha1-96(17), camellia256-cts-cmac(26), camellia128-cts-cmac(25)}) 10.0.184.175: NEEDED_PREAUTH: MYSVC/server.ipa.test@IPA.TEST for krbtgt/IPA.TEST@IPA.TEST, Additional pre-authentication required
Jan 30 04:08:33 server.ipa.test krb5kdc[7795](info): closing down fd 4
Jan 30 04:08:33 server.ipa.test krb5kdc[7793](info): AS_REQ : handle_authdata (2)
Jan 30 04:08:33 server.ipa.test krb5kdc[7793](info): AS_REQ (6 etypes {aes256-cts-hmac-sha384-192(20), aes128-cts-hmac-sha256-128(19), aes256-cts-hmac-sha1-96(18), aes128-cts-hmac-sha1-96(17), camellia256-cts-cmac(26), camellia128-cts-cmac(25)}) 10.0.184.175: HANDLE_AUTHDATA: MYSVC/server.ipa.test@IPA.TEST for krbtgt/IPA.TEST@IPA.TEST, No such file or directory
Jan 30 04:08:33 server.ipa.test krb5kdc[7793](info): closing down fd 4
```
After the fix: can get the TGT and use it.

- [x] use ldapmodify to edit the entry `krbprincipalname=HTTP/server.ipa.test@IPA.TEST,cn=services,cn=accounts,dc=ipa,dc=test` and remove krbcanonicalname, restart IPA, try kinit admin + one ipa * command.
Before the fix:
```
# ipa user-find
ipa: ERROR: No valid Negotiate header in server response
```
After the fix: works well

File: daemons/ipa-slapi-plugins/ipa-sidgen/ipa_sidgen_common.c [hunk 1]
```diff
@@ -554,6 +565,7 @@ int find_sid_for_ldap_entry(struct slapi_entry *entry,
     }
 
 done:
+    slapi_ch_free_string(&uniqueid);
```
Comment (tbordaz) [inline daemons/ipa-slapi-plugins/ipa-sidgen/ipa_sidgen_common.c:568]: I think it is likely invalid as 'uniqueid' value is a reference in the slapi_entry. It should not be free, it is the free of the entry that will free it.
Risk of double free.

Comment (abbra) [reply daemons/ipa-slapi-plugins/ipa-sidgen/ipa_sidgen_common.c:568]: `slapi_entry_attr_get_charptr()` returns malloced memory and explicitly asks for freeing it:
```
char *
slapi_entry_attr_get_charptr(const Slapi_Entry *e, const char *type)
{
    char *p = NULL;
    Slapi_Attr *attr = NULL;

    slapi_entry_attr_find(e, type, &attr);
    if (attr != NULL) {
        Slapi_Value *v;
        const struct berval *bvp;

        slapi_valueset_first_value(&attr->a_present_values, &v);
        bvp = slapi_value_get_berval(v);
        p = slapi_ch_malloc(bvp->bv_len + 1);
        memcpy(p, bvp->bv_val, bvp->bv_len);
        p[bvp->bv_len] = '\0';
    }
    return p;
}
```

Comment (tbordaz) [reply daemons/ipa-slapi-plugins/ipa-sidgen/ipa_sidgen_common.c:568]: oppss you are right. Sorry for the noise

**Comment by flo-renaud** [comment]
master:

* f8dcd78873cc098d5a60e2c56ea4102009631fd6 sidgen: ignore staged users when generating SIDs
* 89d945fe6f9265c5667e825554b2663cc63db3e3 sidgen: fix missing prototypes
* ed977a6e8206366a33fe90ba97844834068f56c8 kdb: PAC generator: do not fail if canonical principal is missing

**Pull request closed without merging**